### PR TITLE
[Data Plane Mgr] Fix issues from test scenario 4.5

### PR DESCRIPTION
**Pull request by cj-chung**

### CODE REVIEW

**Comment by yanmo96** [comment]
Test scenario 4.5 passed with this PR on kolla env 👍

**Comment by yanmo96** [comment]
Verified Test scenario 4 and 5 with this PR on kolla env. Successfully passed 👍

**Comment by xieus** [comment]
@cj-chung @yanmo96 Is this fix verified in the Median cluster?

**Comment by cj-chung** [comment]
As I know, @yanmo96  verified it in kolla-ansible for scenario 4, 4.5, and 5.
@yanmo96 Can you also verify it on Medina cluster?

**Comment by xieus** [comment]
> As I know, @yanmo96 verified it in kolla-ansible for scenario 4, 4.5, and 5.
> @yanmo96 Can you also verify it on Medina cluster?

Just saw Yan's previous comments. As long as we've verified the fix in one cluster, that should be fine.  Thanks guys @cj-chung  @yanmo96

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java [hunk 1]
```diff
@@ -408,6 +491,10 @@ private void patchGoalstateForNeighbor(NetworkConfiguration networkConfig, Unica
             }
         }
 
+        if (patchRouterSubnetStates) {
+            patchGoalstateForRouterSubnet(networkConfig, hostsSubnets, multicastGoalState);
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java:495]: Pls make a comment pointing to the Issue page for tracking purpose.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java [hunk 2]
```diff
@@ -217,6 +217,79 @@ private void patchGoalstateForNeighbor(NetworkConfiguration networkConfig, Unica
...
+                    if (multicastGoalState.getGoalStateBuilder().getSubnetStatesList().stream()
+                            .filter(e -> e.getConfiguration().getId().equals(subnetEntity.getSubnetId()))
+                            .findFirst().orElse(null) == null) {
+                        Subnet.SubnetConfiguration.Builder subnetConfigBuilder = Subnet.SubnetConfiguration.newBuilder();
+                        subnetConfigBuilder.setRevisionNumber(FORMAT_REVISION_NUMBER);
+                        subnetConfigBuilder.setId(subnetEntity.getSubnetId());
+                        subnetConfigBuilder.setVpcId(subnetEntity.getVpcId());
+                        subnetConfigBuilder.setName(subnetEntity.getName());
+                        subnetConfigBuilder.setCidr(subnetEntity.getCidr());
+                        subnetConfigBuilder.setTunnelId(subnetEntity.getTunnelId());
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java:234]: If we use builder in this way, every time we add a new field, we will have to come here and add it accordingly. Could we abstract the conversion logics in a single method, then we only need to change the implementation inside the conversion method? @cj-chung

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java [hunk 3]
```diff
@@ -217,6 +217,79 @@ private void patchGoalstateForNeighbor(NetworkConfiguration networkConfig, Unica
...
+                                if (routerId != null) break;
+                            }
+                        }
+                        Router.RouterConfiguration.Builder routerConfigBuilder = Router.RouterConfiguration.newBuilder();
+                        routerConfigBuilder.setRevisionNumber(FORMAT_REVISION_NUMBER);
+                        routerConfigBuilder.setHostDvrMacAddress(HOST_DVR_MAC);
+                        routerConfigBuilder.setId(routerId);
+                        routerConfigBuilder.addAllSubnetRoutingTables(subnetRoutingTablesList);
+                        Router.RouterState.Builder routerStateBuilder = Router.RouterState.newBuilder();
+                        routerStateBuilder.setConfiguration(routerConfigBuilder.build());
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java:285]: Same comment to Router builder.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java (hunk 1 above)
Comment (cj-chung) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java:460]: comment added

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java (hunk 2 above)
Comment (cj-chung) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java]: abstract the code in this section to `constructSubnetState`.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java (hunk 3 above)
Comment (cj-chung) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java]: abstract the code in this section to `constructRouterState`

**Comment by yanmo96** [comment]
Tested scenario 4, 5, and 4.5 on Medina cluster. All passed! 👍

**Review by xieus — APPROVED** [review]

**Pull request merged**